### DISCUSSION TITLE
Long HLS synthesis time for FC layer with the folding of 1

### DISCUSSION THREAD

**Discussion by haoxi1999** [Xilinx/finn#557] [Q&A] [answered]
Hi, I'm a newcomer to FINN hardware generation. I had some experience in using FINN to transform and extract the parameters of binarized MLPs. I have recently started learning the end-2-end flow of the FINN framework. In my experiment, I attempt to create the hardware with a folding factor of 1 for a 1-bit precision FC layer with 128 neurons and 784-bit input (28*28 pixels from the binarized(bipolar) MNIST dataset). The FINN tool generates the HLS code of StreamingFCLayer_Batch_0 component with MW1=784, MH1=128, PE=128, SIMD=784. However, Vivado HLS seems taking forever synthesise the HLS project. I really appreciate insights into the cause of this long synthesis time and suggestions for improvements.

Thanks!

PS Here's the Vivado HLS log file attached
[vivado_hls.log](https://github.com/Xilinx/finn/files/8113417/vivado_hls.log)

**Comment by mmrahorovic** [comment, marked as answer]
Hi, great to hear you're getting your hands dirty with FINN :) 

I could try my best to help you out with this. I suspect from the log file that you are using the MVAU with 'const' memory mode? Since the weights are binarized and since the weight matrix is not that long, I'm not entirely sure whether the following is applicable. But, it has been observed before that synthesizing the MVAU with the weights 'baked-in' results in long synthesis time (and consumes a lot of memory). See this page for more details: https://finn.readthedocs.io/en/latest/internals.html#streamingfclayer-mem-mode

You could try to set the `mem_mode` to `decoupled` for the MVAU: https://github.com/Xilinx/finn/blob/d1cc9cf94f1c33354cc169c5a6517314d0e94e3b/src/finn/custom_op/fpgadataflow/streamingfclayer_batch.py#L94

I hope this resolves your issue!

**Reply by haoxi1999** [reply]
Thanks Mirza!  I'll try the 'decoupled' memory mode to see if it speeds up the C synthesis. BTW, I have noticed that the HLS synthesis time is significantly shorter when using the feature/vitis_hls or feature/vitis_hls_2021.2 branches with Vivado/Vitis tools 2021.2. Would it be possible that this long HLS synthesis time is only observed in Vivado HLS 2020.1 or specific to HLS codebase for Vivado HLS?

**Reply by mmrahorovic** [reply]
Hi @haoxi1999 ,

You're right indeed, Vitis HLS (2021.2) can potentially lower the HLS synthesis time; the branch `feature/vitis_hls_2021.2` is currently under development and will be ready in one of the upcoming FINN releases, stay tuned :)
For the `decoupled`  memory mode for the `StreamingFCLayer_Batch`, the synthesis time is, to the best of my knowledge, also long with Vitis HLS 2021.2 unfortunately.

Hope your issue got resolved!
